src/java.base/share/classes/com/sun/crypto/provider/KeyWrapCipher.java

```diff
@@ -0,0 +1,753 @@
+        if (!decrypting && dataIdx == 0) {
+            writeIvSemiBlock(dataBuf);
+        }
+        store(in, inOffset, inLen);
+        return null;
+    }
+
+    /**
+     * See CipherSpi.engineUpdate(...) - buffers data internally as
+     * only single part operation is supported.
+     *
+     * @param in the input buffer.
+     * @param inOffset the offset in <code>in</code> where the input
+     * starts.
+     * @param inLen the input length.
+     * @param out the buffer for the result.
+     * @param outOffset the offset in <code>out</code> where the result
+     * is stored.
+     *
+     * @return n/a.
+     *
+     * @exception IllegalStateException upon invocation of this method.
+     */
+    @Override
+    protected int engineUpdate(byte[] in, int inOffset, int inLen,
+            byte[] out, int outOffset) throws ShortBufferException {
+
+        if (!decrypting && dataIdx == 0) {
+            writeIvSemiBlock(dataBuf);
+        }
+
+        store(in, inOffset, inLen);
+        return 0;
+    }
+
+    /**
+     * See CipherSpi.engineDoFinal(...)
+     *
+     * @param input the input buffer
+     * @param inputOffset the offset in <code>in</code> where the input
+     * starts
+     * @param inputLen the input length.
+     *
+     * @return n/a.
+     *
+     * @exception IllegalStateException upon invocation of this method.
+     */
+    @Override
+    protected byte[] engineDoFinal(byte[] in, int inOfs, int inLen)
+            throws IllegalBlockSizeException, BadPaddingException {
+        if (!decrypting && dataIdx == 0) {
+            writeIvSemiBlock(dataBuf);
+        }
+
+        if (inLen > 0) {
+            store(in, inOfs, inLen);
+        }
+
+        // if enc, add padding
+        if (!decrypting) {
+            int actualLen = dataIdx - SEMI_BLKSIZE;
+            if (padding != null) {
+                int paddingLen = padding.padLength(actualLen);
+                // check and re-size dataBuf if needed
+                store(null, 0, paddingLen);
+                try {
+                    padding.padWithLen(dataBuf, dataIdx, paddingLen);
+                    dataIdx += paddingLen;
+                } catch (ShortBufferException sbe) {
+                    // should never happen
+                    throw new AssertionError();
+                }
+            }
+        }
+        try {
+            int outLen;
+            if (decrypting) {
+                outLen = cipher.decryptFinal(dataBuf, 0, dataIdx, null, -1);
+                // unpad if padding is used
+                if (padding != null) {
+                    int padIdx = padding.unpad(dataBuf, 0, outLen);
+                    if (padIdx <= 0) {
+                        throw new BadPaddingException("Bad Padding: " + padIdx);
+                    }
+                    outLen = padIdx;
+                }
+            } else {
+                outLen = cipher.encryptFinal(dataBuf, 0, dataIdx, null, -1);
+            }
```

Comment:
Can we extract this shared logic (among the different versions of `engineDoFinal`) into a separate helper method so that the different `engineDoFinal` methods just need to handle their specific differences (such as returning `Arrays.copyOf(dataBuf, outLen)` or calling `System.arraycopy(dataBuf, 0, out, outOfs, outLen); return outLen;`).